cf_units/_udunits2_parser/__init__.py

```diff
@@ -15,10 +15,217 @@
+        return result
+
+    TERM_HANDLERS = {
+            'CLOSE_PAREN': None,
+            'DATE': str,
+            'DIVIDE': graph.Operand('/'),  # Drop context, such as " per ".
+            'E_POWER': str,
+            'FLOAT': graph.Number,  # Preserve precision as str.
+            'HOUR_MINUTE_SECOND': str,
+            'HOUR_MINUTE': str,
+            'ID': graph.Identifier,
+            'INT': lambda c: graph.Number(int(c)),
+            'MULTIPLY': graph.Operand('*'),
+            'OPEN_PAREN': None,
+            'PERIOD': str,
+            'RAISE': graph.Operand,
+            'TIMESTAMP': graph.Timestamp,
+            'SIGNED_INT': lambda c: graph.Number(int(c)),
+            'SHIFT_OP': None,
+            'WS': None,
+            'UNICODE_EXPONENT': handle_UNICODE_EXPONENT,
+    }
+
+    def visitTerminal(self, ctx):
+        content = ctx.getText()
+
+        symbol_idx = ctx.symbol.type
+        if symbol_idx == -1:
+            # EOF, and all unmatched characters (which will have
+            # already raised a SyntaxError).
+            result = None
+
+        else:
+            name = TOKEN_ID_NAMES[symbol_idx]
+            handler = self.TERM_HANDLERS[name]
+
+            if callable(handler):
+                result = handler(content)
+            else:
+                result = handler
+
+        if result is None:
+            return None
+        elif not isinstance(result, graph.Node):
+            result = graph.Terminal(result)
+        return result
+
+    def visitProduct(self, ctx):
+        # UDUNITS grammar makes no parse distinction for Product
+        # types ('/' and '*'), so we have to do the grunt work here.
+        nodes = self.visitChildren(ctx)
+
+        op_type = graph.Multiply
+
+        if isinstance(nodes, list):
+            last = nodes[-1]
+
+            # Walk the nodes backwards applying the appropriate binary
+            # operation to each node successively.
+            # e.g. 1*2/3*4*5 = 1*(2/(3*(4*5)))
+            for node in nodes[:-1][::-1]:
+                if isinstance(node, graph.Operand):
+                    if node.content == '/':
+                        op_type = graph.Divide
+                    else:
+                        op_type = graph.Multiply
+                else:
+                    last = op_type(node, last)
+            node = last
+        else:
+            node = nodes
+        return node
+
+    def visitTimestamp(self, ctx):
+        # For now, we simply amalgamate timestamps into a single Terminal.
+        # More work is needed to turn this into a good date/time/timezone
+        # representation.
+        return graph.Terminal(ctx.getText())
+
+    def visitPower(self, ctx):
+        node = self.visitChildren(ctx)
+        if isinstance(node, list):
+            if len(node) == 3:
+                # node[1] is the operator, so ignore it.
+                node = graph.Raise(node[0], node[2])
+            else:
+                node = graph.Raise(*node)
```

Comment:
@pelson So this *must* be one or two nodes, right? given the parser grammar - otherwise I assume that ANTLR will catch this for us for free...

Reply:
Confirmed.